src/Aquifer.Common/Services/CachingLanguageService.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Aquifer.Data;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Aquifer.Common.Services;
+
+public interface ICachingLanguageService
+{
+    Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct);
+    Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct);
+    Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct);
+    Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(CancellationToken ct);
+}
+
+public class CachingLanguageService(AquiferDbContext _dbContext, IMemoryCache _memoryCache) : ICachingLanguageService
+{
+    public async Task<string?> GetLanguageCodeAsync(int languageId, CancellationToken ct)
+    {
+        return (await GetLanguageCodeByIdMapAsync(ct))
+            .TryGetValue(languageId, out var languageCode)
+                ? languageCode
+                : null;
+    }
+
+    public async Task<IReadOnlyDictionary<int, string>> GetLanguageCodeByIdMapAsync(CancellationToken ct)
+    {
+        return (await GetDictionariesFromCacheAsync(ct))
+            .LanguageCodeByIdMap;
+    }
+
+    public async Task<int?> GetLanguageIdAsync(string languageCode, CancellationToken ct)
+    {
+        ArgumentNullException.ThrowIfNull(languageCode);
+
+        return (await GetLanguageIdByCodeMapAsync(ct))
+            .TryGetValue(languageCode, out var languageId)
+                ? languageId
+                : null;
+    }
+
+    public async Task<IReadOnlyDictionary<string, int>> GetLanguageIdByCodeMapAsync(CancellationToken ct)
+    {
+        return (await GetDictionariesFromCacheAsync(ct))
+            .LanguageIdByCodeMap;
+    }
+
+    private async Task<(IReadOnlyDictionary<int, string> LanguageCodeByIdMap, IReadOnlyDictionary<string, int> LanguageIdByCodeMap)> GetDictionariesFromCacheAsync(CancellationToken ct)
+    {
+        return await _memoryCache.GetOrCreateAsync(
+            LanguageDictionariesCacheKey,
+            async cacheEntry =>
+            {
+                cacheEntry.SlidingExpiration = s_cacheLifetime;
+
+                var languageIdentifiers = await GetLanguageIdentifiersFromDatabaseAsync(_dbContext, ct);
+
+                return
+                    (LanguageCodeById: languageIdentifiers.ToDictionary(li => li.Id, li => li.Code),
+                    LanguageIdByCode: languageIdentifiers.ToDictionary(li => li.Code, li => li.Id));
+            });
+
+        static async Task<IReadOnlyList<LanguageIdentifier>> GetLanguageIdentifiersFromDatabaseAsync(AquiferDbContext dbContext, CancellationToken ct)
+        {
+            return await dbContext.Languages
+                .Select(l => new LanguageIdentifier(l.Id, l.ISO6393Code))
+                .ToListAsync(ct);
+        }
```

Comment:
Also, not sure if this does anything extra for the performance, but a possible simplification:

![image](https://github.com/user-attachments/assets/49b08fe1-711f-494e-a393-94db603caa94)

That gets rid of the record, and makes for one fewer dictionary mapping.

Reply:
I found it's a helpful practice to separate out the DB logic from the caching logic while still keeping them close together.  Local static functions work well for that because I can give the function a clear name which helps with readability, there's no closure over the enclosing method's variables (because its `static`), and the local function can only be called by the enclosing method (unlike a `private static` method).

I've also seen a repository layer implemented where the repository layer doesn't do caching and thus the caching sits on top of it.  If we ever go that direction then this code would be easier to split out.  Performance wise the DB method call shouldn't happen much (because we're caching the result) so I tend toward separating concerns and readability.

That said, in this case the DB call is very straightforward and this does add more lines of code so I see an argument to inline it.

Reply:
> a possible simplification

> That gets rid of the record, and makes for one fewer dictionary mapping.

Yeah, the additional record was due to the notion that one day we might want to split out the DB logic into a repository layer.

Do you feel strongly that it should be simplified?